### PR TITLE
feat: Expose globbing code from Semaphore to Python

### PR DESCRIPTION
**Pull request by mitsuhiko**
This exposes the globbing code we have in Semaphore through the Python library to Python.

This fulfills two purposes: they are now compatible with each other behavior wise and they both use a non backtracking regex engine which we need for untrusted inputs that we have from rules.

### CODE REVIEW

File: common/src/glob.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    test_glob!("hello.py", "*.py", true, {});
+    test_glob!("hello.py", "*.js", false, {});
+    test_glob!("foo/hello.py", "*.py", true, {});
+    test_glob!("foo/hello.py", "*.py", false, {double_star: true});
+    test_glob!("foo/hello.py", "**/*.py", true, {double_star: true});
+    test_glob!("foo/hello.PY", "**/*.py", false, {double_star: true});
+    test_glob!("foo/hello.PY", "**/*.py", true, {double_star: true, case_insensitive: true});
+    test_glob!("foo\\hello\\bar.PY", "foo/**/*.py", false, {double_star: true, case_insensitive: true});
+    test_glob!("foo\\hello\\bar.PY", "foo/**/*.py", true, {double_star: true, case_insensitive: true, path_normalize: true});
```
Comment (mitsuhiko) [inline common/src/glob.rs:110]: There is a difference in behavior to python here: `**/*.py` actually matches on `foo\\bar.py` even without path normalization. The reason being that a trailing `**` is optional.

I think this is a reasonable change and will not cause any problems. This is already the behavior we have in relay and this makes this consistent now.

File: cabi/src/processing.rs [hunk 2]
```diff
@@ -120,3 +121,32 @@ ffi_fn! {
         panic!("this is a test panic")
     }
 }
+
+/// Controls the globbing behavios
```
Comment (jan-auer) [inline cabi/src/processing.rs]:
```suggestion
/// Controls the globbing behaviors.
```

File: common/src/glob.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+/// Controls the options of the globber.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct GlobOptions {
+    /// When enabled `**` matches over path separators and `*` does not.
+    pub double_star: bool,
+    /// Enables case insensitive path matching.
+    pub case_insensitive: bool,
+    /// Enables path normalization
```
Comment (jan-auer) [inline common/src/glob.rs]:
```suggestion
    /// Enables path normalization.
```

File: common/src/glob.rs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+use globset::GlobBuilder;
+use lazy_static::lazy_static;
+use lru::LruCache;
+use parking_lot::Mutex;
+use regex::bytes::Regex;
+
+lazy_static! {
+    static ref GLOB_CACHE: Mutex<LruCache<(GlobOptions, String), Regex>> =
+        Mutex::new(LruCache::new(500));
```
Comment (jan-auer) [inline common/src/glob.rs:11]: This number is most likely off. I'm pretty sure that with 500 globs, we'll have only evictions in practice, especially if the plan is to use this across all of Sentry.

File: common/src/glob.rs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    test_glob!("hello.py", "*.js", false, {});
+    test_glob!("foo/hello.py", "*.py", true, {});
+    test_glob!("foo/hello.py", "*.py", false, {double_star: true});
+    test_glob!("foo/hello.py", "**/*.py", true, {double_star: true});
+    test_glob!("foo/hello.PY", "**/*.py", false, {double_star: true});
+    test_glob!("foo/hello.PY", "**/*.py", true, {double_star: true, case_insensitive: true});
+    test_glob!("foo\\hello\\bar.PY", "foo/**/*.py", false, {double_star: true, case_insensitive: true});
+    test_glob!("foo\\hello\\bar.PY", "foo/**/*.py", true, {double_star: true, case_insensitive: true, path_normalize: true});
+
+    let mut long_string = std::iter::repeat('x').take(1000000).collect::<String>();
```
Comment (jan-auer) [inline common/src/glob.rs]: You're great at ignoring the linter.

```suggestion
    let mut long_string = std::iter::repeat('x').take(1_000_000).collect::<String>();
```

**Review by jan-auer — APPROVED** [review]
Neat! Only nitpicks

File: common/src/glob.rs (hunk 4 above)
Comment (mitsuhiko) [reply common/src/glob.rs:11]: that's the current number :P

File: cabi/include/semaphore.h [hunk 6]
```diff
@@ -8,6 +8,16 @@
 #include <stdint.h>
 #include <stdlib.h>
 
+/**
+ * Controls the globbing behavios
```
Comment (jan-auer) [inline cabi/include/semaphore.h]: ;)

**Review by jan-auer — APPROVED** [review]

**Pull request merged**